### PR TITLE
routing+discovery: fail non-maxHTLC channel updates in validation

### LINKED ISSUE #7273: [bug]: Channels are disconnected after an "unknown message" error
### Background

Channels with core lightning nodes (most recent 22.11 version and master) frequently disconnect after an "unknown message" error

### Your environment

* version of `lnd` 0.15.5-beta
* 5.15.81 1-NixOS x86_64
* bitcoind v.24.01

### Steps to reproduce

`
Dec 25 00:06:55 lnd[538311]: 2022-12-25 00:06:55.363 [INF] PEER: Peer(02e8f0717f412f455bbcafabc7119f3e3bc1e6c7efeff3df14c98f4da14278dc95): unable to read message from peer: unable to parse message of unknown>
Dec 25 00:06:55 lnd[538311]: 2022-12-25 00:06:55.364 [INF] PEER: Peer(02e8f0717f412f455bbcafabc7119f3e3bc1e6c7efeff3df14c98f4da14278dc95): unable to read message from peer: EOF
Dec 25 00:06:55 lnd[538311]: 2022-12-25 00:06:55.364 [INF] PEER: Peer(02e8f0717f412f455bbcafabc7119f3e3bc1e6c7efeff3df14c98f4da14278dc95): disconnecting 02e8f0717f412f455bbcafabc7119f3e3bc1e6c7efeff3df14c98f>
Dec 25 00:06:55 lnd[538311]: 2022-12-25 00:06:55.364 [INF] NTFN: Cancelling epoch notification, epoch_id=6259
`

This happens frequently, perhaps several times in an hour.

### Expected behaviour

Not disconnecting, or providing more information about the parse error so that it might be investigated

### Actual behaviour

Peers frequently disconnect. I cannot exclude the possibility that some of them are LND peers but so far I have identified two of them as core lightning nodes

### LINKED ISSUE #7290: [bug]: htlc_maximum_msat is not optional anymore in channel updates
### Background

lnd forwards channel updates without htlc_maximum_msat which is now forbidden after https://github.com/lightning/bolts/pull/999 was merged.
The problem is that in this case, other nodes (cln for instance) send back a warning message which triggers a disconnect.

### Your environment

lnd v0.15.5-beta or master

### Steps to reproduce

Run lnd and create a channel to a node running cln.

### Expected behaviour

No channel update messages without htlc_maximum_msat

### Actual behaviour

Some  channel update messages without htlc_maximum_msat. Then a warning is issued by the other node (cln for instance). Then the peer is disconnected in lnd v0.15.5 because the warning message is "unknown" and in master because a warning triggers a disconnect for an active channel.

### PR DESCRIPTION
**Pull request by bitromortac**
## Change Description
Fixes #7290. We don't broadcast or forward any channel updates that are considered invalid by the spec (see change https://github.com/lightning/bolts/commit/6fee63fc342736a2eb7f6e856ae0d85807cc50ec), channel updates that don't set the `must_be_one` (or `option_channel_htlc_max`) flag. It may also fix #7273.

There are several places where we handle channel updates (indicated by @Crypt-iQ in https://github.com/lightningnetwork/lnd/issues/7290#issuecomment-1410927051).

Flows that are handled by this PR are:
* channel funding: `addToRouterGraph`
* receiving and broadcasting channel updates from a peer: `ProcessRemoteAnnouncement`
* updating local channel policies: `PropagateChanPolicyUpdate`

The local graph may still contain older channel updates that don't set the flag. In order to not send those in a historical sync, we check them with `ValidateOptionalFields`. A migration to re-sign possible local legacy channel updates is not straightforward as it would require signing capabilities at migration time. Removing remote channel updates with a migration could be another way to fix this.

Should I rename the `ChanUpdateOptionMaxHtlc` field since it's required and is now named `must_be_one` in the spec?

### CODE REVIEW

**Comment by Crypt-iQ** [comment]
This is very close - not sure why CI isn't running. I think it makes sense to have a commit that renames `ChanUpdateOptionMaxHtlc` as well as `ValidateOptionalFields` since the bit is no longer optional. 

The change to `ValidateOptionalFields` ensures LND doesn't relay updates without the bit because:
- `handleChanUpdate` calls the function before adding the edge or returning it as a gossip message to relay:
https://github.com/lightningnetwork/lnd/blob/4d7ac827732773501ea0643c3c2dc14c6c90f154/discovery/gossiper.go#L2764
- `UpdatesInHorizon` now uses the function in the second commit
- when processing an onion-wrapped error from a node in the route, we'll call `ValidateChannelUpdateAnn` before adding the edge to the graph:
https://github.com/lightningnetwork/lnd/blob/0eb07c944c0c08411621e0112dec7d3f9b0c09f3/routing/router.go#L2423

The first two points mean that we won't relay these updates and will just drop them on the floor. The third point means that an onion-encrypted update won't get into our graph. Note that `addToRouterGraph` in current `master` will always set the bit when sending the update to the gossiper.

File: discovery/chan_series.go [hunk 1]
```diff
@@ -131,10 +132,24 @@ func (c *ChanSeries) UpdatesInHorizon(chain chainhash.Hash,
 
 		updates = append(updates, chanAnn)
 		if edge1 != nil {
-			updates = append(updates, edge1)
+			// We don't want to send channel updates that don't
+			// conform to the spec (anymore).
+			err := routing.ValidateChannelUpdateFields(0, edge1)
+			if err != nil {
+				log.Debugf("Invalid channel update %v: %v",
```
Comment (yyforyongyu) [inline discovery/chan_series.go:139]: Maybe this should be an error log?

**Comment by bitromortac** [comment]
I tested the following regtest setup: `A (patched LND - sends invalid update) - B (LND - this PR) - C (core-lightning)`
Node B detects the invalid update and does not relay it anymore.

File: discovery/chan_series.go (hunk 1 above)
Comment (bitromortac) [reply discovery/chan_series.go]: Changed it to an error log, I was not really sure here if we should surface it by default

**Review by Crypt-iQ — APPROVED** [review]

**Review by ellemouton — COMMENTED** [review]
Looking good! Nice and simple diff 🔥

**Comment by ellemouton** [comment]
qq @Crypt-iQ:

> when processing an onion-wrapped error from a node in the route, we'll call ValidateChannelUpdateAnn before adding the edge to the graph which means that an onion-encrypted update won't get into our graph.

I think there is a chance that we add an invalid update to our graph, then update LND to include this change and restart. Then there is a (small) chance that we call [this](https://github.com/lightningnetwork/lnd/blob/5d22d5eb6c7877ff2f6f082f764f3b5ae37a2eef/htlcswitch/link.go#L660-L683) with an invalid update. right?

Should we perhaps just validate the update there too?

File: channeldb/migration/lnwire21/channel_update.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ type ChanUpdateMsgFlags uint8
 const (
 	// ChanUpdateOptionMaxHtlc is a bit that indicates whether the
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply channeldb/migration/lnwire21/channel_update.go]: ah yes, thank you!

File: discovery/gossiper_test.go [hunk 3]
```diff
@@ -2921,8 +2941,8 @@ func TestOptionalFieldsChannelUpdateValidation(t *testing.T) {
 		t.Fatalf("expected chan update to error, instead got %v", err)
 	}
 
-	// The final update should succeed, since setting the flag 0 means the
-	// nonsense max_htlc field will just be ignored.
+	// The final update should not succeed, a channel update with no message
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply discovery/gossiper_test.go]: I fully agree, added :+1:

**Review by ellemouton — APPROVED** [review]
Noice Noice Noice 🔥

**Pull request merged**